--- psplash/psplash.c~	2008-08-14 14:58:01.000000000 +0200
+++ psplash/psplash.c	2008-08-14 14:58:01.000000000 +0200
@@ -49,7 +49,7 @@
 			fb->height - (fb->height/6) - h, 
 			fb->width,
 			h,
-			0xec, 0xec, 0xe1);
+			0xff, 0xff, 0xff);
 
   psplash_fb_draw_text (fb,
 			(fb->width-w)/2, 
@@ -75,7 +75,7 @@
       barwidth = (CLAMP(value,0,100) * width) / 100;
       psplash_fb_draw_rect (fb, x + barwidth, y, 
     			width - barwidth, height,
-			0xec, 0xec, 0xe1);
+			0xec, 0xec, 0xe1);
       psplash_fb_draw_rect (fb, x, y, barwidth,
 			    height, 0x6d, 0x6d, 0x70);
     }
@@ -84,7 +84,7 @@
       barwidth = (CLAMP(-value,0,100) * width) / 100;
       psplash_fb_draw_rect (fb, x, y, 
     			width - barwidth, height,
-			0xec, 0xec, 0xe1);
+			0xec, 0xec, 0xe1);
       psplash_fb_draw_rect (fb, x + width - barwidth,
 			    y, barwidth, height,
 			    0x6d, 0x6d, 0x70);
@@ -253,7 +253,7 @@
     exit(-1);
 
   /* Clear the background with #ecece1 */
-  psplash_fb_draw_rect (fb, 0, 0, fb->width, fb->height, 0xec, 0xec, 0xe1);
+  psplash_fb_draw_rect (fb, 0, 0, fb->width, fb->height, 0xff, 0xff, 0xff);
 
   /* Draw the OH logo  */
   psplash_fb_draw_image (fb, 
